syntax = "proto3";

package raftpb;


enum raftState {
    Follower = 0;
    Candidate = 1;
    Leader = 2;
}

message ReqVote {
    uint64 term = 1;
    uint32 candidateId = 2;
    uint64 lastLogTerm = 3;
    uint64 lastLogIndex = 4;
}

message RespVote {
    uint64 term = 1;
    bool voteGranted = 2;
}

message Entry {
    uint64 term = 1;
    uint64 index = 2;
    bytes record = 3;
}

message ReqAppendEntry {
    uint64 term = 1;
    uint32 leaderId = 2;
    uint64 prevLogTerm = 3;
    uint64 prevLogIndex = 4;
    repeated Entry entries = 5;
    uint64 leaderCommitIndex = 6;
}

message RespAppendEntry {
    uint64 term = 1;
    bool success = 2;
}
message ReqTran {
    uint32 src = 1;
    bytes record = 2;
}

message RespTran {
    uint32 err = 1;
    string errmsg = 2;
}

service raftService {
    rpc RequestVote(ReqVote) returns(RespVote);
    rpc AppendEntries(ReqAppendEntry) returns (RespAppendEntry);
    rpc Tran(ReqTran) returns (RespTran);
}